File: examples/api-key-management/web2/csharp/api-key-management-api/Controllers/ApiKeyController.cs
using Microsoft.AspNetCore.Mvc;
using System.Security.Cryptography;
using System.Text;
using System.Collections.Concurrent;

namespace api_key_management_api.Controllers;

[ApiController]
[Route("api")]
public class ApiKeyController : ControllerBase
{
    // Mock DB: SHA256 Hash -> Metadata
    private static readonly ConcurrentDictionary<string, KeyMetadata> _keysDb = new();

    public class KeyMetadata
    {
        public string UserId { get; set; } = string.Empty;
        public string Plan { get; set; } = "basic";
        public bool IsActive { get; set; } = true;
        public int UsageCount { get; set; } = 0;
    }

    public class CreateKeyRequest { public string UserId { get; set; } = string.Empty; public string Plan { get; set; } = "basic"; }

    private static string HashKey(string key)
    {
        byte[] bytes = SHA256.HashData(Encoding.UTF8.GetBytes(key));
        return Convert.ToHexString(bytes).ToLower();
    }

    [HttpPost("keys")]
    public IActionResult CreateKey([FromBody] CreateKeyRequest request)
    {
        if (string.IsNullOrEmpty(request.UserId)) return BadRequest(new { error = "UserId required" });

        string rawKey = $"sk_test_{Guid.NewGuid():n}";
        string hashedKey = HashKey(rawKey);

        _keysDb[hashedKey] = new KeyMetadata { UserId = request.UserId, Plan = request.Plan };

        return Created("", new { message = "API Key created. Save it now.", apiKey = rawKey });
    }

    [HttpGet("protected-data")]
    public IActionResult GetProtectedData()
    {
        string? authHeader = Request.Headers.Authorization;
        if (string.IsNullOrEmpty(authHeader) || !authHeader.StartsWith("Bearer "))
            return Unauthorized(new { error = "Missing or invalid Authorization header" });

        string token = authHeader["Bearer ".Length..];
        string hashedToken = HashKey(token);

        if (!_keysDb.TryGetValue(hashedToken, out var meta) || !meta.IsActive)
            return Unauthorized(new { error = "Invalid or revoked key" });

        meta.UsageCount++;
        return Ok(new { 
            message = "Access granted to .NET API!", 
            userId = meta.UserId, 
            data = "Highly sensitive C# records" 
        });
    }
}

File: examples/api-key-management/web2/springboot/src/main/java/com/example/demo/ApiKeyController.java
package com.example.demo;

import org.springframework.web.bind.annotation.*;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@RestController
@RequestMapping("/api")
public class ApiKeyController {

    // Mock DB: Hash -> KeyMetadata
    private static final Map<String, KeyMetadata> apiKeysDb = new ConcurrentHashMap<>();

    // Record for Key Metadata
    public static class KeyMetadata {
        public String userId;
        public String plan;
        public boolean isActive;
        public int usageCount;

        public KeyMetadata(String userId, String plan) {
            this.userId = userId;
            this.plan = plan;
            this.isActive = true;
            this.usageCount = 0;
        }
    }

    // Response object for key creation
    public static class CreateKeyResponse {
        public String message;
        public String apiKey;

        public CreateKeyResponse(String message, String apiKey) {
            this.message = message;
            this.apiKey = apiKey;
        }
    }

    private String hashKey(String key) throws NoSuchAlgorithmException {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] encodedHash = digest.digest(key.getBytes());
        StringBuilder hexString = new StringBuilder(2 * encodedHash.length);
        for (byte b : encodedHash) {
            String hex = Integer.toHexString(0xff & b);
            if (hex.length() == 1) hexString.append('0');
            hexString.append(hex);
        }
        return hexString.toString();
    }

    @PostMapping("/keys")
    public ResponseEntity<?> createKey(@RequestBody Map<String, String> payload) throws NoSuchAlgorithmException {
        String userId = payload.get("user_id");
        if (userId == null) return ResponseEntity.badRequest().body("user_id is required");

        String rawKey = "sk_test_" + UUID.randomUUID().toString().replace("-", "");
        String hashedKey = hashKey(rawKey);

        apiKeysDb.put(hashedKey, new KeyMetadata(userId, payload.getOrDefault("plan", "basic")));

        return new ResponseEntity<>(new CreateKeyResponse("Key generated. Store safely.", rawKey), HttpStatus.CREATED);
    }

    @GetMapping("/protected-data")
    public ResponseEntity<?> getProtectedData(@RequestHeader(value = "Authorization", required = false) String authHeader) throws NoSuchAlgorithmException {
        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Missing Authorization header");
        }

        String token = authHeader.substring(7);
        String hashedToken = hashKey(token);

        KeyMetadata meta = apiKeysDb.get(hashedToken);
        if (meta == null || !meta.isActive) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("Invalid or revoked API Key");
        }

        meta.usageCount++;
        return ResponseEntity.ok(Map.of(
            "message", "Access granted to Spring Boot!",
            "user_id", meta.userId,
            "data", "Confidential enterprise report"
        ));
    }
}
